File: 04-linked-lists/08_reorder_list.c
#include "../test_runner.h"

typedef struct ListNode {
    int val;
    struct ListNode* next;
} ListNode;

ListNode* make_node(int val) {
    ListNode* n = malloc(sizeof(ListNode));
    n->val = val; n->next = NULL;
    return n;
}

ListNode* make_list(int* arr, int n) {
    if (n == 0) return NULL;
    ListNode* head = make_node(arr[0]);
    ListNode* curr = head;
    for (int i = 1; i < n; i++) {
        curr->next = make_node(arr[i]);
        curr = curr->next;
    }
    return head;
}

void to_array(ListNode* head, int* out, int* len) {
    *len = 0;
    while (head) { out[(*len)++] = head->val; head = head->next; }
}


void reorder_list(ListNode* head) {
    /* your code */
}

int main() {
    int out[10], len;
    ListNode* h = make_list(ARR(1,2,3,4), 4);
    reorder_list(h);
    to_array(h, out, &len);
    TEST_ARR("4 nodes", out, ARR(1,4,2,3), 4);

    h = make_list(ARR(1,2,3,4,5), 5);
    reorder_list(h);
    to_array(h, out, &len);
    TEST_ARR("5 nodes", out, ARR(1,5,2,4,3), 5);
    TEST_SUMMARY();
}

File: 04-linked-lists/08_reorder_list.py
import sys, os; sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), '..'))
from test_runner import run_tests


# Reorder List (LC #143) -- Medium
# Reorder: L0 -> Ln -> L1 -> Ln-1 -> ...
#
#   Key: Find middle, reverse second half, merge alternating

class ListNode:
    def __init__(self, val=0, nxt=None):
        self.val = val
        self.next = nxt


def make_list(arr):
    dummy = ListNode(0)
    curr = dummy
    for v in arr:
        curr.next = ListNode(v)
        curr = curr.next
    return dummy.next


def to_array(head):
    arr = []
    while head:
        arr.append(head.val)
        head = head.next
    return arr


def reorder_list(head):
    """Reorder in-place: L0->Ln->L1->Ln-1->..."""
    pass


def solve(arr):
    head = make_list(arr)
    reorder_list(head)
    return to_array(head)


run_tests(solve, [
    (([1,2,3,4],),   [1,4,2,3]),
    (([1,2,3,4,5],), [1,5,2,4,3]),
    (([1],),         [1]),
    (([1,2],),       [1,2]),
])
